src/load_file.cpp

```diff
@@ -1,17 +1,52 @@
+/*******************************************************************************
+ *
+ * MIT License
+ *
+ * Copyright (c) 2021 Advanced Micro Devices, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ *
+ *******************************************************************************/
+
+#include <miopen/errors.hpp>
 #include <miopen/load_file.hpp>
-#include <sstream>
 #include <fstream>
+#include <ios>
+#include <iterator>
+#include <vector>
 
 namespace miopen {
 
-std::string LoadFile(const fs::path& p) { return LoadFile(p.string()); }
-
-std::string LoadFile(const std::string& s)
+std::vector<char> LoadFile(const fs::path& path)
 {
-    const std::ifstream t(s);
-    std::stringstream buffer;
-    buffer << t.rdbuf();
-    return buffer.str();
+    std::ifstream in(path, std::ios::binary | std::ios::ate);
+    if(!in.is_open())
+        MIOPEN_THROW("Unable to open file: " + path.string());
+    in.unsetf(std::ios::skipws);
+    auto size = in.tellg();
+    in.seekg(0, std::ios::beg);
```

Comment:
I prefer to keep `tellg()`—the file is already opened, and with `file_size`, on Linux, uses `stat` to read `st_size`, but Windows uses the `GetFileSize()` function, which requires opening the file first. The size is only required to estimate roughly how much memory we need to preallocate. It doesn't have to be precise, but for the binary stream, `tellg()` returns the number of bytes from the beginning of the file (both Linux and Windows). It is not valid for other stream modes on Windows, so I prefer to keep a standard and seemingly deterministic version for both operating systems.

Reply:
I'm no fully convinced.
`GetFileSize()` requires file handle, which can be just an identifier to get all the file information from the filesystem, while `std::ifstream in(path, std::ios::binary | std::ios::ate);` really opens the file and rewinds its internal state to the end of the file (which can be potentially zero-cost, but who knows). Then you need to rewind file back (which is also can be zero-cost, but so `GetFileSize()` can be). Morever, MSDN says:

> You can use SetFilePointerEx to determine the length of a file. To do this, use FILE_END for dwMoveMethod and seek to location zero. The file offset returned is the length of the file. However, this practice can have unintended side effects, such as failure to save the current file pointer so that the program can return to that location. It is simpler and safer to use the [GetFileSizeEx](https://learn.microsoft.com/en-us/windows/desktop/api/fileapi/nf-fileapi-getfilesizeex) function instead.

Nevertheless the code with rewinding looks unclear.

Reply:
I ran quick performance measurements on Linux and Windows with `gtell()` and `file_size()`. The test reads 100 times the file of size 400MB. The results revealed that both methods are equally performant (approximately 10ms in favor of `file_size`).
However, I discovered that determining file size is not an issue here. The actual slowdown is copying data from file to vector. I changed to allocate the whole vector and used `in.read()`. I observed x3 faster file reading on Linux and x7 faster file reading on Windows. With that change, Windows reads files faster than Linux (for the mentioned test, ~300ms for Linux and ~225ms for Windows per single file read).